src/v/security/authorizer.h

```diff
@@ -290,21 +292,111 @@ class authorizer final {
+              !role
+              || *role != nullptr && (*role)->type() == principal_type::role);
+            const acl_principal_base& to_check = *role.value_or(&user);
+            bool is_allow = perm == acl_permission::allow;
+            std::optional<acl_matches::acl_match> entry;
+            if (is_allow) {
+                entry = acl_any_implied_ops_allowed(
+                  acls, to_check, host, operation);
+            } else {
+                entry = acls.find(operation, to_check, host, perm);
+            }
+            if (!entry) {
+                return std::nullopt;
+            }
+            switch (to_check.type()) {
+            case principal_type::user:
+            case principal_type::ephemeral_user:
+                return auth_result::acl_match(
+                  user, host, operation, resource_name, is_allow, *entry);
+            case principal_type::role:
+                return auth_result::role_acl_match(
+                  user,
+                  security::role_name{to_check.name_view()},
+                  host,
+                  operation,
+                  resource_name,
+                  is_allow,
+                  *entry);
+            }
+            __builtin_unreachable();
+        };
+
+        auto check_deny =
+          [&check_access](
+            const security::acl_principal& user,
+            std::optional<const security::acl_principal_view*> role
+            = std::nullopt) {
+              return check_access(user, role, acl_permission::deny);
+          };
+
+        auto check_allow =
+          [&check_access](
+            const security::acl_principal& user,
+            std::optional<const security::acl_principal_view*> role
+            = std::nullopt) {
+              return check_access(user, role, acl_permission::allow);
+          };
+
+        if (auto result = check_deny(principal); result.has_value()) {
+            return std::move(result.value());
+        }
+
+        auto member_roles =
+          [this, &principal]() -> security::role_store::roles_range {
+            if (principal.type() == security::principal_type::user) {
+                return _role_store->roles_for_member(
+                  security::role_member_view::from_principal(principal));
+            } else {
+                return {};
+            }
+        }();
+
+        auto access_view = [&member_roles](
+                             const auto& fn, const acl_principal& user) {
+            return member_roles | std::views::transform([](const auto& e) {
+                       return role::to_principal_view(e);
+                   })
+                   | std::views::transform(
+                     [&user, &fn](const auto& e) { return fn(user, &e); })
+                   | std::views::filter(
+                     [](const std::optional<auth_result>& r) {
+                         return r.has_value();
+                     })
+                   | std::views::take(1);
```

Comment:
nitpick: It might be worth returning an optional here to keep the same pattern as `check_access`.

Reply:
done